{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "399e1507",
   "metadata": {},
   "source": [
    "Purpose: Prepare data for XGBoost regression.<br>\n",
    "Author: Anna Pardo<br>\n",
    "Date initiated: Dec. 19, 2025"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "08af1976",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from sklearn.feature_selection import VarianceThreshold\n",
    "from sklearn.preprocessing import StandardScaler\n",
    "from functools import reduce"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "dda61c38",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_256/2025937544.py:2: DtypeWarning: Columns (1) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  mdtpm = pd.read_csv(\"~/Yucca_genomics/rna_insilico_genome/TPM/Yg_toYgIS_allTPM_correctedmd_over1mil.txt\",sep=\"\\t\",header=\"infer\")\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_name</th>\n",
       "      <th>genotype</th>\n",
       "      <th>time</th>\n",
       "      <th>treat</th>\n",
       "      <th>ZT</th>\n",
       "      <th>species</th>\n",
       "      <th>Yucal.01G000100.v2.1</th>\n",
       "      <th>Yucal.01G000200.v2.1</th>\n",
       "      <th>Yucal.01G000300.v2.1</th>\n",
       "      <th>Yucal.01G000400.v2.1</th>\n",
       "      <th>...</th>\n",
       "      <th>YufilH1095122m.g</th>\n",
       "      <th>YufilH1095123m.g</th>\n",
       "      <th>YufilH1095125m.g</th>\n",
       "      <th>YufilH1095126m.g</th>\n",
       "      <th>YufilH1095128m.g</th>\n",
       "      <th>YufilH1095131m.g</th>\n",
       "      <th>YufilH1095132m.g</th>\n",
       "      <th>YufilH1095134m.g</th>\n",
       "      <th>YufilH1095146m.g</th>\n",
       "      <th>YufilH1095147m.g</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Y1</td>\n",
       "      <td>18</td>\n",
       "      <td>1.0</td>\n",
       "      <td>W</td>\n",
       "      <td>1.0</td>\n",
       "      <td>gloriosa</td>\n",
       "      <td>34.002815</td>\n",
       "      <td>4.546167</td>\n",
       "      <td>0.0</td>\n",
       "      <td>17.236119</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>6.167441</td>\n",
       "      <td>1.435253</td>\n",
       "      <td>0.279077</td>\n",
       "      <td>11.420139</td>\n",
       "      <td>0.662252</td>\n",
       "      <td>1.886709</td>\n",
       "      <td>6.675105</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Y10</td>\n",
       "      <td>2AB</td>\n",
       "      <td>1.5</td>\n",
       "      <td>W</td>\n",
       "      <td>3.0</td>\n",
       "      <td>gloriosa</td>\n",
       "      <td>40.070758</td>\n",
       "      <td>3.628454</td>\n",
       "      <td>0.0</td>\n",
       "      <td>14.918115</td>\n",
       "      <td>...</td>\n",
       "      <td>0.713535</td>\n",
       "      <td>2.197522</td>\n",
       "      <td>15.695904</td>\n",
       "      <td>1.193256</td>\n",
       "      <td>10.172780</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.214483</td>\n",
       "      <td>12.125756</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Y100</td>\n",
       "      <td>2AB</td>\n",
       "      <td>6.5</td>\n",
       "      <td>W</td>\n",
       "      <td>23.0</td>\n",
       "      <td>gloriosa</td>\n",
       "      <td>47.402599</td>\n",
       "      <td>5.201760</td>\n",
       "      <td>0.0</td>\n",
       "      <td>17.406497</td>\n",
       "      <td>...</td>\n",
       "      <td>0.314746</td>\n",
       "      <td>2.261806</td>\n",
       "      <td>21.742515</td>\n",
       "      <td>1.798380</td>\n",
       "      <td>10.256681</td>\n",
       "      <td>1.748663</td>\n",
       "      <td>2.075757</td>\n",
       "      <td>25.215640</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.838780</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Y101</td>\n",
       "      <td>2AB</td>\n",
       "      <td>1.0</td>\n",
       "      <td>D</td>\n",
       "      <td>1.0</td>\n",
       "      <td>gloriosa</td>\n",
       "      <td>57.062380</td>\n",
       "      <td>6.374324</td>\n",
       "      <td>0.0</td>\n",
       "      <td>10.567561</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.072366</td>\n",
       "      <td>27.573939</td>\n",
       "      <td>1.164591</td>\n",
       "      <td>9.105769</td>\n",
       "      <td>1.257431</td>\n",
       "      <td>2.431447</td>\n",
       "      <td>4.508369</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.813682</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Y103</td>\n",
       "      <td>2AB</td>\n",
       "      <td>1.0</td>\n",
       "      <td>D</td>\n",
       "      <td>1.0</td>\n",
       "      <td>gloriosa</td>\n",
       "      <td>34.679279</td>\n",
       "      <td>6.087451</td>\n",
       "      <td>0.0</td>\n",
       "      <td>11.115252</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.914550</td>\n",
       "      <td>28.621412</td>\n",
       "      <td>0.869052</td>\n",
       "      <td>7.076224</td>\n",
       "      <td>0.515567</td>\n",
       "      <td>2.419222</td>\n",
       "      <td>4.625880</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.867418</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 85968 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  sample_name genotype  time treat    ZT   species  Yucal.01G000100.v2.1  \\\n",
       "0          Y1       18   1.0     W   1.0  gloriosa             34.002815   \n",
       "1         Y10      2AB   1.5     W   3.0  gloriosa             40.070758   \n",
       "2        Y100      2AB   6.5     W  23.0  gloriosa             47.402599   \n",
       "3        Y101      2AB   1.0     D   1.0  gloriosa             57.062380   \n",
       "4        Y103      2AB   1.0     D   1.0  gloriosa             34.679279   \n",
       "\n",
       "   Yucal.01G000200.v2.1  Yucal.01G000300.v2.1  Yucal.01G000400.v2.1  ...  \\\n",
       "0              4.546167                   0.0             17.236119  ...   \n",
       "1              3.628454                   0.0             14.918115  ...   \n",
       "2              5.201760                   0.0             17.406497  ...   \n",
       "3              6.374324                   0.0             10.567561  ...   \n",
       "4              6.087451                   0.0             11.115252  ...   \n",
       "\n",
       "   YufilH1095122m.g  YufilH1095123m.g  YufilH1095125m.g  YufilH1095126m.g  \\\n",
       "0          0.000000          6.167441          1.435253          0.279077   \n",
       "1          0.713535          2.197522         15.695904          1.193256   \n",
       "2          0.314746          2.261806         21.742515          1.798380   \n",
       "3          0.000000          1.072366         27.573939          1.164591   \n",
       "4          0.000000          0.914550         28.621412          0.869052   \n",
       "\n",
       "   YufilH1095128m.g  YufilH1095131m.g  YufilH1095132m.g  YufilH1095134m.g  \\\n",
       "0         11.420139          0.662252          1.886709          6.675105   \n",
       "1         10.172780          0.000000          2.214483         12.125756   \n",
       "2         10.256681          1.748663          2.075757         25.215640   \n",
       "3          9.105769          1.257431          2.431447          4.508369   \n",
       "4          7.076224          0.515567          2.419222          4.625880   \n",
       "\n",
       "   YufilH1095146m.g  YufilH1095147m.g  \n",
       "0               0.0          0.000000  \n",
       "1               0.0          0.000000  \n",
       "2               0.0          1.838780  \n",
       "3               0.0          0.813682  \n",
       "4               0.0          0.867418  \n",
       "\n",
       "[5 rows x 85968 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# load Yg TPM\n",
    "mdtpm = pd.read_csv(\"~/Yucca_genomics/rna_insilico_genome/TPM/Yg_toYgIS_allTPM_correctedmd_over1mil.txt\",sep=\"\\t\",header=\"infer\")\n",
    "mdtpm.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "18505147",
   "metadata": {},
   "outputs": [],
   "source": [
    "mdtpm.drop([\"time\",\"species\"],axis=1,inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "fcfb513c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['18', '2AB', '1AB', '19', '15', 'Eudy', 'G', '56', '36', '13',\n",
       "       '45', '52', '43', '37', '48', '55', '70', '61', '51', '46', '53',\n",
       "       '16', '6', '12', '20', '50'], dtype=object)"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mdtpm[\"genotype\"] = mdtpm[\"genotype\"].astype(str)\n",
    "mdtpm[\"genotype\"].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "3633767f",
   "metadata": {},
   "outputs": [],
   "source": [
    "zttpm = mdtpm[mdtpm[\"ZT\"].isin([1.0,5.0,9.0,13.0,17.0,21.0])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "d704544d",
   "metadata": {},
   "outputs": [],
   "source": [
    "mdonly = zttpm[[\"sample_name\",\"genotype\",\"treat\",\"ZT\"]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "a374ed2d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# check how many and which genotypes have complete data (at least 1 replicate in each treatment+ZT)\n",
    "def iscomplete(gt):\n",
    "    gtmd = mdonly[mdonly[\"genotype\"]==gt]\n",
    "    gtcount = gtmd.groupby([\"treat\",\"ZT\"]).count().reset_index()[[\"treat\",\"ZT\",\"sample_name\"]].rename(columns={\"sample_name\":\"nreps\"})\n",
    "    if len(gtcount.index)==12:\n",
    "        return \"complete\"\n",
    "    elif len(gtcount.index)<12:\n",
    "        return \"incomplete\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "ee6aadd9",
   "metadata": {},
   "outputs": [],
   "source": [
    "dfdict={\"Genotype\":[],\"Complete\":[]}\n",
    "for g in mdonly[\"genotype\"].unique():\n",
    "    dfdict['Genotype'].append(g)\n",
    "    dfdict['Complete'].append(iscomplete(g))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "7e0629c0",
   "metadata": {},
   "outputs": [],
   "source": [
    "cdf = pd.DataFrame(dfdict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "d3a16a83",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Genotype</th>\n",
       "      <th>Complete</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>15</td>\n",
       "      <td>incomplete</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>52</td>\n",
       "      <td>incomplete</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>55</td>\n",
       "      <td>incomplete</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>70</td>\n",
       "      <td>incomplete</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>53</td>\n",
       "      <td>incomplete</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>16</td>\n",
       "      <td>incomplete</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>20</td>\n",
       "      <td>incomplete</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>50</td>\n",
       "      <td>incomplete</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Genotype    Complete\n",
       "4        15  incomplete\n",
       "11       52  incomplete\n",
       "15       55  incomplete\n",
       "16       70  incomplete\n",
       "20       53  incomplete\n",
       "21       16  incomplete\n",
       "24       20  incomplete\n",
       "25       50  incomplete"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cdf[cdf[\"Complete\"]==\"incomplete\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "0f59cb0b",
   "metadata": {},
   "outputs": [],
   "source": [
    "def iscompletedf(gt):\n",
    "    gtmd = mdonly[mdonly[\"genotype\"]==gt]\n",
    "    gtcount = gtmd.groupby([\"treat\",\"ZT\"]).count().reset_index()[[\"treat\",\"ZT\",\"sample_name\"]].rename(columns={\"sample_name\":\"nreps\"})\n",
    "    return gtcount"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "e0b60ddc",
   "metadata": {},
   "outputs": [],
   "source": [
    "complete_gts = list(cdf[cdf[\"Complete\"]==\"complete\"][\"Genotype\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "caf879dd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "18"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(complete_gts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "3a121c50",
   "metadata": {},
   "outputs": [],
   "source": [
    "comptpm = zttpm[zttpm[\"genotype\"].isin(complete_gts)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "a7856006",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['15', '52', '55', '70', '53', '16', '20', '50']"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list(cdf[cdf[\"Complete\"]==\"incomplete\"][\"Genotype\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "705d5c33",
   "metadata": {},
   "outputs": [],
   "source": [
    "# for comptpm: filter out genes that are expressed in <10% of genotypes (i.e., expressed only in 1 or 2 genotypes)\n",
    "def egtnum(gid):\n",
    "    tpmsub = comptpm[[\"genotype\",gid]]\n",
    "    maxtpm = tpmsub.groupby(\"genotype\").max().reset_index()\n",
    "    # count the number of genotypes with max>0\n",
    "    max0 = maxtpm[maxtpm[gid]>0]\n",
    "    return len(max0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "8c9a1c64",
   "metadata": {},
   "outputs": [],
   "source": [
    "dropgenes = []\n",
    "for c in comptpm.columns:\n",
    "    if c.startswith(\"Y\"):\n",
    "        if egtnum(c)<=2:\n",
    "            dropgenes.append(c)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "73b74585",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "4143"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(dropgenes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "1e002d69",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "85962"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(comptpm.columns)-4"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "b7820a69",
   "metadata": {},
   "outputs": [],
   "source": [
    "filtcomp = comptpm.drop(dropgenes,axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "77e2d767",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "81819"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(filtcomp.columns)-4"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "83b58170",
   "metadata": {},
   "outputs": [],
   "source": [
    "# drop zero-variance genes\n",
    "def variance_threshold_selector(data):\n",
    "    selector = VarianceThreshold()\n",
    "    selector.fit(data)\n",
    "    return data[data.columns[selector.get_support(indices=True)]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "6dd9508c",
   "metadata": {},
   "outputs": [],
   "source": [
    "filttpm = filtcomp.set_index([\"sample_name\",\"genotype\",\"treat\",\"ZT\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "cfe63457",
   "metadata": {},
   "outputs": [],
   "source": [
    "vttpm = variance_threshold_selector(filttpm)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "99cf67f1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of zero-variance genes removed: 0\n"
     ]
    }
   ],
   "source": [
    "print(\"Number of zero-variance genes removed:\",len(filttpm.columns)-len(vttpm.columns))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "612bc1ad",
   "metadata": {},
   "outputs": [],
   "source": [
    "vttpm_log = vttpm.apply(lambda x: np.log2(x+1))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "93b6af32",
   "metadata": {},
   "outputs": [],
   "source": [
    "vttpm_log.reset_index(inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "16b65f8b",
   "metadata": {},
   "outputs": [],
   "source": [
    "vttpm_log.drop(\"sample_name\",axis=1,inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "972319ec",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genotype</th>\n",
       "      <th>treat</th>\n",
       "      <th>ZT</th>\n",
       "      <th>Yucal.01G000100.v2.1</th>\n",
       "      <th>Yucal.01G000200.v2.1</th>\n",
       "      <th>Yucal.01G000300.v2.1</th>\n",
       "      <th>Yucal.01G000400.v2.1</th>\n",
       "      <th>Yucal.01G000500.v2.1</th>\n",
       "      <th>Yucal.01G000600.v2.1</th>\n",
       "      <th>Yucal.01G000700.v2.1</th>\n",
       "      <th>...</th>\n",
       "      <th>YufilH1095122m.g</th>\n",
       "      <th>YufilH1095123m.g</th>\n",
       "      <th>YufilH1095125m.g</th>\n",
       "      <th>YufilH1095126m.g</th>\n",
       "      <th>YufilH1095128m.g</th>\n",
       "      <th>YufilH1095131m.g</th>\n",
       "      <th>YufilH1095132m.g</th>\n",
       "      <th>YufilH1095134m.g</th>\n",
       "      <th>YufilH1095147m.g</th>\n",
       "      <th>treat_ZT</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>18</td>\n",
       "      <td>W</td>\n",
       "      <td>1.0</td>\n",
       "      <td>5.129399</td>\n",
       "      <td>2.471491</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.188727</td>\n",
       "      <td>2.140948</td>\n",
       "      <td>3.181674</td>\n",
       "      <td>3.168583</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.841458</td>\n",
       "      <td>1.284072</td>\n",
       "      <td>0.355103</td>\n",
       "      <td>3.634609</td>\n",
       "      <td>0.733139</td>\n",
       "      <td>1.529426</td>\n",
       "      <td>2.940186</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>W_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2AB</td>\n",
       "      <td>D</td>\n",
       "      <td>1.0</td>\n",
       "      <td>5.859532</td>\n",
       "      <td>2.882511</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.532013</td>\n",
       "      <td>2.045877</td>\n",
       "      <td>3.441016</td>\n",
       "      <td>2.950223</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.051279</td>\n",
       "      <td>4.836628</td>\n",
       "      <td>1.114095</td>\n",
       "      <td>3.337107</td>\n",
       "      <td>1.174682</td>\n",
       "      <td>1.778817</td>\n",
       "      <td>2.461625</td>\n",
       "      <td>0.858921</td>\n",
       "      <td>D_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2AB</td>\n",
       "      <td>D</td>\n",
       "      <td>1.0</td>\n",
       "      <td>5.157015</td>\n",
       "      <td>2.825267</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.598753</td>\n",
       "      <td>2.070301</td>\n",
       "      <td>3.257171</td>\n",
       "      <td>2.777274</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.937005</td>\n",
       "      <td>4.888569</td>\n",
       "      <td>0.902307</td>\n",
       "      <td>3.013681</td>\n",
       "      <td>0.599857</td>\n",
       "      <td>1.773668</td>\n",
       "      <td>2.492079</td>\n",
       "      <td>0.901045</td>\n",
       "      <td>D_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1AB</td>\n",
       "      <td>W</td>\n",
       "      <td>21.0</td>\n",
       "      <td>5.791442</td>\n",
       "      <td>2.753920</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.225544</td>\n",
       "      <td>1.127059</td>\n",
       "      <td>2.864516</td>\n",
       "      <td>2.795164</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.217595</td>\n",
       "      <td>4.454795</td>\n",
       "      <td>1.659643</td>\n",
       "      <td>3.064140</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.809274</td>\n",
       "      <td>4.059519</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>W_21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>18</td>\n",
       "      <td>W</td>\n",
       "      <td>21.0</td>\n",
       "      <td>5.596679</td>\n",
       "      <td>3.241637</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.047620</td>\n",
       "      <td>1.632024</td>\n",
       "      <td>3.407077</td>\n",
       "      <td>3.284910</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.086022</td>\n",
       "      <td>1.656996</td>\n",
       "      <td>2.910847</td>\n",
       "      <td>3.955409</td>\n",
       "      <td>1.919294</td>\n",
       "      <td>1.736134</td>\n",
       "      <td>4.441396</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>W_21</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 81823 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  genotype treat    ZT  Yucal.01G000100.v2.1  Yucal.01G000200.v2.1  \\\n",
       "0       18     W   1.0              5.129399              2.471491   \n",
       "1      2AB     D   1.0              5.859532              2.882511   \n",
       "2      2AB     D   1.0              5.157015              2.825267   \n",
       "3      1AB     W  21.0              5.791442              2.753920   \n",
       "4       18     W  21.0              5.596679              3.241637   \n",
       "\n",
       "   Yucal.01G000300.v2.1  Yucal.01G000400.v2.1  Yucal.01G000500.v2.1  \\\n",
       "0                   0.0              4.188727              2.140948   \n",
       "1                   0.0              3.532013              2.045877   \n",
       "2                   0.0              3.598753              2.070301   \n",
       "3                   0.0              4.225544              1.127059   \n",
       "4                   0.0              4.047620              1.632024   \n",
       "\n",
       "   Yucal.01G000600.v2.1  Yucal.01G000700.v2.1  ...  YufilH1095122m.g  \\\n",
       "0              3.181674              3.168583  ...               0.0   \n",
       "1              3.441016              2.950223  ...               0.0   \n",
       "2              3.257171              2.777274  ...               0.0   \n",
       "3              2.864516              2.795164  ...               0.0   \n",
       "4              3.407077              3.284910  ...               0.0   \n",
       "\n",
       "   YufilH1095123m.g  YufilH1095125m.g  YufilH1095126m.g  YufilH1095128m.g  \\\n",
       "0          2.841458          1.284072          0.355103          3.634609   \n",
       "1          1.051279          4.836628          1.114095          3.337107   \n",
       "2          0.937005          4.888569          0.902307          3.013681   \n",
       "3          1.217595          4.454795          1.659643          3.064140   \n",
       "4          3.086022          1.656996          2.910847          3.955409   \n",
       "\n",
       "   YufilH1095131m.g  YufilH1095132m.g  YufilH1095134m.g  YufilH1095147m.g  \\\n",
       "0          0.733139          1.529426          2.940186          0.000000   \n",
       "1          1.174682          1.778817          2.461625          0.858921   \n",
       "2          0.599857          1.773668          2.492079          0.901045   \n",
       "3          0.000000          1.809274          4.059519          0.000000   \n",
       "4          1.919294          1.736134          4.441396          0.000000   \n",
       "\n",
       "   treat_ZT  \n",
       "0       W_1  \n",
       "1       D_1  \n",
       "2       D_1  \n",
       "3      W_21  \n",
       "4      W_21  \n",
       "\n",
       "[5 rows x 81823 columns]"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "vttpm_log[\"treat_ZT\"] = vttpm_log[\"treat\"]+\"_\"+vttpm_log[\"ZT\"].astype(int).astype(str)\n",
    "vttpm_log.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "afe4ca9f",
   "metadata": {},
   "outputs": [],
   "source": [
    "vttpm_log.drop([\"treat\",\"ZT\"],axis=1,inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "f8848d4e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# calculate mean expression of each gene on a per-treatment/ZT basis\n",
    "## define a function to do this\n",
    "def genemeans(gid):\n",
    "    sublog = vttpm_log[[\"genotype\",\"treat_ZT\",gid]]\n",
    "    sublog.groupby([\"genotype\",\"treat_ZT\"]).mean().reset_index()\n",
    "    slp = sublog.pivot_table(index=\"genotype\",columns=\"treat_ZT\",values=gid).reset_index()\n",
    "    for c in slp.columns:\n",
    "        if c!=\"genotype\":\n",
    "            slp.rename(columns={c:gid+\"_\"+c},inplace=True)\n",
    "    return slp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "abc6c8de",
   "metadata": {},
   "outputs": [],
   "source": [
    "dflist = []\n",
    "for c in vttpm_log.columns:\n",
    "    if c.startswith(\"Y\"):\n",
    "        dflist.append(genemeans(c))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "69b96b42",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>treat_ZT</th>\n",
       "      <th>genotype</th>\n",
       "      <th>Yucal.01G000100.v2.1_D_1</th>\n",
       "      <th>Yucal.01G000100.v2.1_D_13</th>\n",
       "      <th>Yucal.01G000100.v2.1_D_17</th>\n",
       "      <th>Yucal.01G000100.v2.1_D_21</th>\n",
       "      <th>Yucal.01G000100.v2.1_D_5</th>\n",
       "      <th>Yucal.01G000100.v2.1_D_9</th>\n",
       "      <th>Yucal.01G000100.v2.1_W_1</th>\n",
       "      <th>Yucal.01G000100.v2.1_W_13</th>\n",
       "      <th>Yucal.01G000100.v2.1_W_17</th>\n",
       "      <th>Yucal.01G000100.v2.1_W_21</th>\n",
       "      <th>Yucal.01G000100.v2.1_W_5</th>\n",
       "      <th>Yucal.01G000100.v2.1_W_9</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>12</td>\n",
       "      <td>4.671265</td>\n",
       "      <td>4.921500</td>\n",
       "      <td>5.005791</td>\n",
       "      <td>4.798222</td>\n",
       "      <td>4.095183</td>\n",
       "      <td>4.459452</td>\n",
       "      <td>4.871347</td>\n",
       "      <td>4.712149</td>\n",
       "      <td>4.941031</td>\n",
       "      <td>5.006829</td>\n",
       "      <td>4.102163</td>\n",
       "      <td>4.456983</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>13</td>\n",
       "      <td>5.144395</td>\n",
       "      <td>4.975804</td>\n",
       "      <td>4.853688</td>\n",
       "      <td>5.012945</td>\n",
       "      <td>4.893604</td>\n",
       "      <td>5.055911</td>\n",
       "      <td>5.258964</td>\n",
       "      <td>5.489978</td>\n",
       "      <td>5.268651</td>\n",
       "      <td>5.200845</td>\n",
       "      <td>4.848472</td>\n",
       "      <td>5.053348</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>18</td>\n",
       "      <td>5.684827</td>\n",
       "      <td>5.435823</td>\n",
       "      <td>5.405006</td>\n",
       "      <td>5.475807</td>\n",
       "      <td>5.751970</td>\n",
       "      <td>5.898819</td>\n",
       "      <td>5.573681</td>\n",
       "      <td>5.574150</td>\n",
       "      <td>5.244135</td>\n",
       "      <td>5.290898</td>\n",
       "      <td>6.075920</td>\n",
       "      <td>5.711747</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>19</td>\n",
       "      <td>5.135665</td>\n",
       "      <td>5.190652</td>\n",
       "      <td>5.463501</td>\n",
       "      <td>5.250987</td>\n",
       "      <td>4.594409</td>\n",
       "      <td>5.076206</td>\n",
       "      <td>5.157064</td>\n",
       "      <td>5.322314</td>\n",
       "      <td>5.205319</td>\n",
       "      <td>4.971768</td>\n",
       "      <td>4.843764</td>\n",
       "      <td>5.160359</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1AB</td>\n",
       "      <td>5.494941</td>\n",
       "      <td>5.422841</td>\n",
       "      <td>5.219464</td>\n",
       "      <td>4.980347</td>\n",
       "      <td>5.043138</td>\n",
       "      <td>5.326056</td>\n",
       "      <td>5.546432</td>\n",
       "      <td>5.651210</td>\n",
       "      <td>5.358873</td>\n",
       "      <td>5.537470</td>\n",
       "      <td>5.627045</td>\n",
       "      <td>5.456193</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "treat_ZT genotype  Yucal.01G000100.v2.1_D_1  Yucal.01G000100.v2.1_D_13  \\\n",
       "0              12                  4.671265                   4.921500   \n",
       "1              13                  5.144395                   4.975804   \n",
       "2              18                  5.684827                   5.435823   \n",
       "3              19                  5.135665                   5.190652   \n",
       "4             1AB                  5.494941                   5.422841   \n",
       "\n",
       "treat_ZT  Yucal.01G000100.v2.1_D_17  Yucal.01G000100.v2.1_D_21  \\\n",
       "0                          5.005791                   4.798222   \n",
       "1                          4.853688                   5.012945   \n",
       "2                          5.405006                   5.475807   \n",
       "3                          5.463501                   5.250987   \n",
       "4                          5.219464                   4.980347   \n",
       "\n",
       "treat_ZT  Yucal.01G000100.v2.1_D_5  Yucal.01G000100.v2.1_D_9  \\\n",
       "0                         4.095183                  4.459452   \n",
       "1                         4.893604                  5.055911   \n",
       "2                         5.751970                  5.898819   \n",
       "3                         4.594409                  5.076206   \n",
       "4                         5.043138                  5.326056   \n",
       "\n",
       "treat_ZT  Yucal.01G000100.v2.1_W_1  Yucal.01G000100.v2.1_W_13  \\\n",
       "0                         4.871347                   4.712149   \n",
       "1                         5.258964                   5.489978   \n",
       "2                         5.573681                   5.574150   \n",
       "3                         5.157064                   5.322314   \n",
       "4                         5.546432                   5.651210   \n",
       "\n",
       "treat_ZT  Yucal.01G000100.v2.1_W_17  Yucal.01G000100.v2.1_W_21  \\\n",
       "0                          4.941031                   5.006829   \n",
       "1                          5.268651                   5.200845   \n",
       "2                          5.244135                   5.290898   \n",
       "3                          5.205319                   4.971768   \n",
       "4                          5.358873                   5.537470   \n",
       "\n",
       "treat_ZT  Yucal.01G000100.v2.1_W_5  Yucal.01G000100.v2.1_W_9  \n",
       "0                         4.102163                  4.456983  \n",
       "1                         4.848472                  5.053348  \n",
       "2                         6.075920                  5.711747  \n",
       "3                         4.843764                  5.160359  \n",
       "4                         5.627045                  5.456193  "
      ]
     },
     "execution_count": 68,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dflist[0].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "6da42257",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "81819"
      ]
     },
     "execution_count": 69,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(dflist)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "24247d25",
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "Input \u001b[0;32mIn [33]\u001b[0m, in \u001b[0;36m<cell line: 1>\u001b[0;34m()\u001b[0m\n\u001b[0;32m----> 1\u001b[0m allfeatures \u001b[38;5;241m=\u001b[39m \u001b[43mreduce\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43;01mlambda\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[43mx\u001b[49m\u001b[43m,\u001b[49m\u001b[43my\u001b[49m\u001b[43m:\u001b[49m\u001b[43m \u001b[49m\u001b[43mpd\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mmerge\u001b[49m\u001b[43m(\u001b[49m\u001b[43mx\u001b[49m\u001b[43m,\u001b[49m\u001b[43my\u001b[49m\u001b[43m,\u001b[49m\u001b[43mon\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mgenotype\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\u001b[43mdflist\u001b[49m\u001b[43m)\u001b[49m\n",
      "Input \u001b[0;32mIn [33]\u001b[0m, in \u001b[0;36m<lambda>\u001b[0;34m(x, y)\u001b[0m\n\u001b[0;32m----> 1\u001b[0m allfeatures \u001b[38;5;241m=\u001b[39m reduce(\u001b[38;5;28;01mlambda\u001b[39;00m x,y: \u001b[43mpd\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mmerge\u001b[49m\u001b[43m(\u001b[49m\u001b[43mx\u001b[49m\u001b[43m,\u001b[49m\u001b[43my\u001b[49m\u001b[43m,\u001b[49m\u001b[43mon\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mgenotype\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m)\u001b[49m,dflist)\n",
      "File \u001b[0;32m~/.local/lib/python3.8/site-packages/pandas/core/reshape/merge.py:122\u001b[0m, in \u001b[0;36mmerge\u001b[0;34m(left, right, how, on, left_on, right_on, left_index, right_index, sort, suffixes, copy, indicator, validate)\u001b[0m\n\u001b[1;32m     90\u001b[0m \u001b[38;5;129m@Substitution\u001b[39m(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;130;01m\\n\u001b[39;00m\u001b[38;5;124mleft : DataFrame or named Series\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m     91\u001b[0m \u001b[38;5;129m@Appender\u001b[39m(_merge_doc, indents\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m0\u001b[39m)\n\u001b[1;32m     92\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mmerge\u001b[39m(\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    105\u001b[0m     validate: \u001b[38;5;28mstr\u001b[39m \u001b[38;5;241m|\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m,\n\u001b[1;32m    106\u001b[0m ) \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m>\u001b[39m DataFrame:\n\u001b[1;32m    107\u001b[0m     op \u001b[38;5;241m=\u001b[39m _MergeOperation(\n\u001b[1;32m    108\u001b[0m         left,\n\u001b[1;32m    109\u001b[0m         right,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    120\u001b[0m         validate\u001b[38;5;241m=\u001b[39mvalidate,\n\u001b[1;32m    121\u001b[0m     )\n\u001b[0;32m--> 122\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mop\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget_result\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/.local/lib/python3.8/site-packages/pandas/core/reshape/merge.py:725\u001b[0m, in \u001b[0;36m_MergeOperation.get_result\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    722\u001b[0m lindexers \u001b[38;5;241m=\u001b[39m {\u001b[38;5;241m1\u001b[39m: left_indexer} \u001b[38;5;28;01mif\u001b[39;00m left_indexer \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m \u001b[38;5;28;01melse\u001b[39;00m {}\n\u001b[1;32m    723\u001b[0m rindexers \u001b[38;5;241m=\u001b[39m {\u001b[38;5;241m1\u001b[39m: right_indexer} \u001b[38;5;28;01mif\u001b[39;00m right_indexer \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m \u001b[38;5;28;01melse\u001b[39;00m {}\n\u001b[0;32m--> 725\u001b[0m result_data \u001b[38;5;241m=\u001b[39m \u001b[43mconcatenate_managers\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    726\u001b[0m \u001b[43m    \u001b[49m\u001b[43m[\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mleft\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_mgr\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mlindexers\u001b[49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mright\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_mgr\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mrindexers\u001b[49m\u001b[43m)\u001b[49m\u001b[43m]\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    727\u001b[0m \u001b[43m    \u001b[49m\u001b[43maxes\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m[\u001b[49m\u001b[43mllabels\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mappend\u001b[49m\u001b[43m(\u001b[49m\u001b[43mrlabels\u001b[49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mjoin_index\u001b[49m\u001b[43m]\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    728\u001b[0m \u001b[43m    \u001b[49m\u001b[43mconcat_axis\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;241;43m0\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m    729\u001b[0m \u001b[43m    \u001b[49m\u001b[43mcopy\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mcopy\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    730\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    732\u001b[0m typ \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mleft\u001b[38;5;241m.\u001b[39m_constructor\n\u001b[1;32m    733\u001b[0m result \u001b[38;5;241m=\u001b[39m typ(result_data)\u001b[38;5;241m.\u001b[39m__finalize__(\u001b[38;5;28mself\u001b[39m, method\u001b[38;5;241m=\u001b[39m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_merge_type)\n",
      "File \u001b[0;32m~/.local/lib/python3.8/site-packages/pandas/core/internals/concat.py:242\u001b[0m, in \u001b[0;36mconcatenate_managers\u001b[0;34m(mgrs_indexers, axes, concat_axis, copy)\u001b[0m\n\u001b[1;32m    240\u001b[0m     fastpath \u001b[38;5;241m=\u001b[39m blk\u001b[38;5;241m.\u001b[39mvalues\u001b[38;5;241m.\u001b[39mdtype \u001b[38;5;241m==\u001b[39m values\u001b[38;5;241m.\u001b[39mdtype\n\u001b[1;32m    241\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m--> 242\u001b[0m     values \u001b[38;5;241m=\u001b[39m \u001b[43m_concatenate_join_units\u001b[49m\u001b[43m(\u001b[49m\u001b[43mjoin_units\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mconcat_axis\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mcopy\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mcopy\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    243\u001b[0m     fastpath \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mFalse\u001b[39;00m\n\u001b[1;32m    245\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m fastpath:\n",
      "File \u001b[0;32m~/.local/lib/python3.8/site-packages/pandas/core/internals/concat.py:545\u001b[0m, in \u001b[0;36m_concatenate_join_units\u001b[0;34m(join_units, concat_axis, copy)\u001b[0m\n\u001b[1;32m    542\u001b[0m has_none_blocks \u001b[38;5;241m=\u001b[39m \u001b[38;5;28many\u001b[39m(unit\u001b[38;5;241m.\u001b[39mblock\u001b[38;5;241m.\u001b[39mdtype\u001b[38;5;241m.\u001b[39mkind \u001b[38;5;241m==\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mV\u001b[39m\u001b[38;5;124m\"\u001b[39m \u001b[38;5;28;01mfor\u001b[39;00m unit \u001b[38;5;129;01min\u001b[39;00m join_units)\n\u001b[1;32m    543\u001b[0m upcasted_na \u001b[38;5;241m=\u001b[39m _dtype_to_na_value(empty_dtype, has_none_blocks)\n\u001b[0;32m--> 545\u001b[0m to_concat \u001b[38;5;241m=\u001b[39m [\n\u001b[1;32m    546\u001b[0m     ju\u001b[38;5;241m.\u001b[39mget_reindexed_values(empty_dtype\u001b[38;5;241m=\u001b[39mempty_dtype, upcasted_na\u001b[38;5;241m=\u001b[39mupcasted_na)\n\u001b[1;32m    547\u001b[0m     \u001b[38;5;28;01mfor\u001b[39;00m ju \u001b[38;5;129;01min\u001b[39;00m join_units\n\u001b[1;32m    548\u001b[0m ]\n\u001b[1;32m    550\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mlen\u001b[39m(to_concat) \u001b[38;5;241m==\u001b[39m \u001b[38;5;241m1\u001b[39m:\n\u001b[1;32m    551\u001b[0m     \u001b[38;5;66;03m# Only one block, nothing to concatenate.\u001b[39;00m\n\u001b[1;32m    552\u001b[0m     concat_values \u001b[38;5;241m=\u001b[39m to_concat[\u001b[38;5;241m0\u001b[39m]\n",
      "File \u001b[0;32m~/.local/lib/python3.8/site-packages/pandas/core/internals/concat.py:546\u001b[0m, in \u001b[0;36m<listcomp>\u001b[0;34m(.0)\u001b[0m\n\u001b[1;32m    542\u001b[0m has_none_blocks \u001b[38;5;241m=\u001b[39m \u001b[38;5;28many\u001b[39m(unit\u001b[38;5;241m.\u001b[39mblock\u001b[38;5;241m.\u001b[39mdtype\u001b[38;5;241m.\u001b[39mkind \u001b[38;5;241m==\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mV\u001b[39m\u001b[38;5;124m\"\u001b[39m \u001b[38;5;28;01mfor\u001b[39;00m unit \u001b[38;5;129;01min\u001b[39;00m join_units)\n\u001b[1;32m    543\u001b[0m upcasted_na \u001b[38;5;241m=\u001b[39m _dtype_to_na_value(empty_dtype, has_none_blocks)\n\u001b[1;32m    545\u001b[0m to_concat \u001b[38;5;241m=\u001b[39m [\n\u001b[0;32m--> 546\u001b[0m     \u001b[43mju\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget_reindexed_values\u001b[49m\u001b[43m(\u001b[49m\u001b[43mempty_dtype\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mempty_dtype\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mupcasted_na\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mupcasted_na\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    547\u001b[0m     \u001b[38;5;28;01mfor\u001b[39;00m ju \u001b[38;5;129;01min\u001b[39;00m join_units\n\u001b[1;32m    548\u001b[0m ]\n\u001b[1;32m    550\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mlen\u001b[39m(to_concat) \u001b[38;5;241m==\u001b[39m \u001b[38;5;241m1\u001b[39m:\n\u001b[1;32m    551\u001b[0m     \u001b[38;5;66;03m# Only one block, nothing to concatenate.\u001b[39;00m\n\u001b[1;32m    552\u001b[0m     concat_values \u001b[38;5;241m=\u001b[39m to_concat[\u001b[38;5;241m0\u001b[39m]\n",
      "File \u001b[0;32m~/.local/lib/python3.8/site-packages/pandas/core/internals/concat.py:525\u001b[0m, in \u001b[0;36mJoinUnit.get_reindexed_values\u001b[0;34m(self, empty_dtype, upcasted_na)\u001b[0m\n\u001b[1;32m    523\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m    524\u001b[0m     \u001b[38;5;28;01mfor\u001b[39;00m ax, indexer \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mindexers\u001b[38;5;241m.\u001b[39mitems():\n\u001b[0;32m--> 525\u001b[0m         values \u001b[38;5;241m=\u001b[39m \u001b[43malgos\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mtake_nd\u001b[49m\u001b[43m(\u001b[49m\u001b[43mvalues\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mindexer\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43maxis\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43max\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    527\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m values\n",
      "File \u001b[0;32m~/.local/lib/python3.8/site-packages/pandas/core/array_algos/take.py:117\u001b[0m, in \u001b[0;36mtake_nd\u001b[0;34m(arr, indexer, axis, fill_value, allow_fill)\u001b[0m\n\u001b[1;32m    114\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m arr\u001b[38;5;241m.\u001b[39mtake(indexer, fill_value\u001b[38;5;241m=\u001b[39mfill_value, allow_fill\u001b[38;5;241m=\u001b[39mallow_fill)\n\u001b[1;32m    116\u001b[0m arr \u001b[38;5;241m=\u001b[39m np\u001b[38;5;241m.\u001b[39masarray(arr)\n\u001b[0;32m--> 117\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43m_take_nd_ndarray\u001b[49m\u001b[43m(\u001b[49m\u001b[43marr\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mindexer\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43maxis\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mfill_value\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mallow_fill\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/.local/lib/python3.8/site-packages/pandas/core/array_algos/take.py:161\u001b[0m, in \u001b[0;36m_take_nd_ndarray\u001b[0;34m(arr, indexer, axis, fill_value, allow_fill)\u001b[0m\n\u001b[1;32m    157\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m    158\u001b[0m     out \u001b[38;5;241m=\u001b[39m np\u001b[38;5;241m.\u001b[39mempty(out_shape, dtype\u001b[38;5;241m=\u001b[39mdtype)\n\u001b[1;32m    160\u001b[0m func \u001b[38;5;241m=\u001b[39m _get_take_nd_function(\n\u001b[0;32m--> 161\u001b[0m     \u001b[43marr\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mndim\u001b[49m, arr\u001b[38;5;241m.\u001b[39mdtype, out\u001b[38;5;241m.\u001b[39mdtype, axis\u001b[38;5;241m=\u001b[39maxis, mask_info\u001b[38;5;241m=\u001b[39mmask_info\n\u001b[1;32m    162\u001b[0m )\n\u001b[1;32m    163\u001b[0m func(arr, indexer, out, fill_value)\n\u001b[1;32m    165\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m flip_order:\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "allfeatures = reduce(lambda x,y: pd.merge(x,y,on=\"genotype\"),dflist)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d7e4c502",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
